chapters/part3/hyperparameter-optimization-with-optuna.md

```diff
@@ -0,0 +1,362 @@
+local dropout = std.parseJson(std.extVar('dropout'));
+local lr = std.parseJson(std.extVar('lr'));
+local max_filter_size = std.parseInt(std.extVar('max_filter_size'));
+local num_filters = std.parseInt(std.extVar('num_filters'));
+local output_dim = std.parseInt(std.extVar('output_dim'));
+local ngram_filter_sizes = std.range(2, max_filter_size);
+
+{
+  numpy_seed: seed,
+  pytorch_seed: seed,
+  random_seed: seed,
+  dataset_reader: {
+    type: 'text_classification_json',
+    tokenizer: {
+      type: 'whitespace',
+    },
+    token_indexers: {
+      tokens: {
+        type: 'single_id',
+      },
+    },
+  },
+  datasets_for_vocab_creation: ['train'],
+  train_data_path: train_data_path,
+  validation_data_path: validation_data_path,
+  model: {
+    type: 'basic_classifier',
+    text_field_embedder: {
+      token_embedders: {
+        tokens: {
+          embedding_dim: embedding_dim,
+        },
+      },
+    },
+    seq2vec_encoder: {
+      type: 'cnn',
+      embedding_dim: embedding_dim,
+      ngram_filter_sizes: ngram_filter_sizes,
+      num_filters: num_filters,
+      output_dim: output_dim,
+    },
+    dropout: dropout,
+  },
+  data_loader: {
+    batch_size: batch_size,
+  },
+  trainer: {
+    cuda_device: cuda_device,
+    num_epochs: num_epochs,
+    optimizer: {
+      lr: lr,
+      type: 'sgd',
+    },
+    validation_metric: '+accuracy',
+  },
+}
+```
+
+</details>
+
+## II: Defining Search Space for Hyperparameters in Pythonic way
+
+Now that you have created the config, the next step is defining the hyperparameter search spaces.
+In Optuna, a search space is defined by creating an `objective function`.
+Each hyperparameter search space is declared with `suggest_int` or `suggest_float`.
+For categorical hyperparameters, you can use `suggest_categorical`.
+Please see <a href="https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.Trial.html#optuna.trial.Trial">our documentation</a> for more information.
+In this tutorial, I create an objective function like following (Note that the names of the hyperparameters should be the same as those defined in the configuration earlier):
+
+```python
+import optuna
+
+
+def objective(trial: optuna.Trial) -> float:
+    trial.suggest_int("embedding_dim", 32, 256)
+    trial.suggest_int("max_filter_size", 2, 6)
+    trial.suggest_int("num_filters", 32, 256)
+    trial.suggest_int("output_dim", 32, 256)
+    trial.suggest_float("dropout", 0.0, 0.8)
+    trial.suggest_float("lr", 5e-3, 5e-1, log=True)
+
+    executor = optuna.integration.allennlp.AllenNLPExecutor(
+        trial=trial,  # trial object
+        config_file="./config/imdb_optuna.jsonnet",  # path to jsonnet
+        serialization_dir=f"./result/optuna/{trial.number}",
+        metrics="best_validation_accuracy"
+    )
+    return executor.run()
+```
+
+After defining search spaces using `trial.suggest_int` and `trial.suggest_float`, the `trial` object should be passed to `AllenNLPExecutor`.
```

Comment:
Maybe we can explain the role of `trial` here something like as follows:

The `trial` object holds all suggested values after defining search spaces using `trial.suggest_int` and `trial.suggest_float`,  and it should be pass to `AllenNLPExecutor`.

Reply:
I edited the PR to use your suggestion. a52b7b3